### DISCUSSION TITLE
Unexpected results for substructure search

### DISCUSSION THREAD

**Discussion by susanhleung** [rdkit/rdkit#7254] [Q&A]
Hi all!

I expect both of these SMARTS queries to match my molecule, does anyone know why the second query (patt2) does not match?

```
In [1]: import rdkit
   ...: from rdkit import Chem
   ...: print(rdkit.__version__)
   ...: from rdkit import Chem
   ...: mol = Chem.MolFromSmiles('NCC1CCCCC1')
   ...: patt1 = Chem.MolFromSmarts('[#6](-[#7&D1])-[#1,#6]')
   ...: patt2 = Chem.MolFromSmarts('[#1,#6]-[#6](-[#7&D1])-[#1,#6]')
   ...: print(mol.GetSubstructMatch(patt1), mol.GetSubstructMatch(patt2))
2023.09.6
(1, 0, 2) ()
```

Many thanks,

Susan

**Comment by bp-kelley** [comment]
Smiles uses implicit hydrogens by default, your query is asking for an explicit hydrogen so can never match.

These style of queries either require recursive smarts, or you can make a new molecule calling Chem.AddHs and do the structure match on that.

**Comment by susanhleung** [comment]
Thanks Brian! Yes makes sense... I didn't think about implicit hydrogens.

I would prefer not to come up with recursive smarts so I tried as suggested with a new molecule using Chem.AddHs but still can't get it to match? Also tried reading in with the removeHs=False option, any ideas?

In [1]: import rdkit
   ...: from rdkit import Chem
   ...: print(rdkit.__version__)
   ...: from rdkit import Chem
   ...: patt2 = Chem.MolFromSmarts('[#1,#6]-[#6](-[#7&D1])-[#1,#6]')
   ...: mol = Chem.MolFromSmiles('NCC1CCCCC1')
   ...: mh = Chem.AddHs(mol)
   ...: print(Chem.MolToSmiles(mh))
   ...: print(mh.GetSubstructMatch(patt2))
   ...: 
   ...: params = Chem.SmilesParserParams()
   ...: params.removeHs = False
   ...: molH = Chem.MolFromSmiles('[H]N([H])C([H])([H])C1([H])C([H])([H])C([H])([H])C([H])([H])C([H])([H])C1([H])[H]', params)
   ...: print(Chem.MolToSmiles(molH))
   ...: print(molH.GetSubstructMatch(patt2))
2023.09.6
[H]N([H])C([H])([H])C1([H])C([H])([H])C([H])([H])C([H])([H])C([H])([H])C1([H])[H]
()
[H]N([H])C([H])([H])C1([H])C([H])([H])C([H])([H])C([H])([H])C([H])([H])C1([H])[H]
()

As an aside, we originally encountered this issue using the PostgreSQL cartridge and using mol_from_ctab to read in the database molecules. However, I believe the hydrogens are also lost when using mol_from_ctab and I can't see a function to addHs to database molecules either unless I overlooked this?

Thanks,

Susan

**Reply by bp-kelley** [reply]
The main problem here is using D which means explicit connections, without Hs `[#6](-[#7&D1])-[#1,#6]` matches the N, but with Hs you need D3.

It might be better to use X which is all connections.

```
from rdkit import Chem
from rdkit import Chem
mol = Chem.MolFromSmiles('NCC1CCCCC1')
patt1 = Chem.MolFromSmarts('[#6](-[#7&X3])-[#1,#6]')
patt2 = Chem.MolFromSmarts('[#1,#6]-[#6](-[#7&X3])-[#1,#6]')
molh = Chem.AddHs(mol)
print(mol.GetSubstructMatch(patt1), mol.GetSubstructMatch(patt2))
print(molh.GetSubstructMatch(patt1), molh.GetSubstructMatch(patt2))
```

```
(1, 0, 2) ()
(1, 0, 2) (2, 1, 0, 10)
```

**Reply by bp-kelley** [reply]
https://www.daylight.com/dayhtml/doc/theory/theory.smarts.html
